setup-cli v2 sometimes completes successfully but supabase is missing in the next step

We’re seeing an intermittent failure with `supabase/setup-cli`.

  Pinned action:

  ```yaml
  uses: supabase/setup-cli@2eca1b4d35d9016e560c763c9159a350aa80c370
```

  Our workflow shape is:

  - uses: supabase/setup-
  cli@2eca1b4d35d9016e560c763c9159a350aa80c370

  - run: supabase start --exclude "edge-
  runtime,imgproxy,logflare,mailpit,postgres-
  meta,realtime,studio,supavisor,vector"

  Failure:

  Run bun src/main.ts
  ...
  Run supabase start --exclude "..."
  /home/runner/_work/_temp/...sh: line 1: supabase: command not
  found
  Error: Process completed with exit code 127

  What’s strange is that setup-cli itself completed successfully.

  I compared this with a successful sibling job from the same GitHub
  Actions run:

  - successful job: next step had
    SUPABASE_INTERNAL_IMAGE_REGISTRY=ghcr.io and supabase start
    worked
  - failing job: next step did not have
    SUPABASE_INTERNAL_IMAGE_REGISTRY, and supabase was not on PATH

  That makes it look like the action sometimes exits green without
  persisting its side effects to the next step.

  Relevant source in the pinned revision:

  if (process.argv[1] === fileURLToPath(import.meta.url)) {
    void run();
  }

  run() is async and does the important work:

  core.addPath(cliPath);
  core.exportVariable(CLI_CONFIG_REGISTRY, "ghcr.io");

  This looks like a plausible race.

  Environment:

  - GitHub Actions
  - self-hosted Blacksmith runner
  - Ubuntu 24.04
  - runner 2.333.1

  Can you confirm whether the async entrypoint should be awaited,
  and whether that could explain this kind of flake?

--- a/src/main.test.ts
+++ b/src/main.test.ts
@@ -167,39 +167,52 @@ async function getMainModule(): Promise<typeof import("./main.ts")> {
   return mainModule;
 }
 
-async function waitForCalls(assertion: () => void): Promise<void> {
-  let failure: Error | null = null;
-
-  for (let attempt = 0; attempt < 50; attempt += 1) {
-    try {
-      assertion();
-      return;
-    } catch (error) {
-      failure = error instanceof Error ? error : new Error(String(error));
-      await Bun.sleep(10);
-    }
-  }
-
-  throw failure ?? new Error("Timed out waiting for action side effects");
-}
-
-test("runs the action entrypoint with omitted version and latest fallback", async () => {
+test("awaits the action entrypoint with omitted version and latest fallback", async () => {
   process.env.GITHUB_WORKSPACE = repo;
   const cliDir = createFakeCli("supabase 2.84.2");
-  const spies = createActionSpies("", cliDir, "/latest/download/");
+  let startDownload!: () => void;
+  let finishDownload!: () => void;
+  const downloadStarted = new Promise<void>((resolve) => {
+    startDownload = resolve;
+  });
+  const downloadFinished = new Promise<string>((resolve) => {
+    finishDownload = () => resolve(path.join(os.tmpdir(), "supabase-cli.tar.gz"));
+  });
+  const spies = {
+    getInput: spyOn(core, "getInput").mockReturnValue(""),
+    setOutput: spyOn(core, "setOutput").mockImplementation(() => {}),
+    addPath: spyOn(core, "addPath").mockImplementation(() => {}),
+    exportVariable: spyOn(core, "exportVariable").mockImplementation(() => {}),
+    setFailed: spyOn(core, "setFailed").mockImplementation(() => {}),
+    downloadTool: spyOn(tc, "downloadTool").mockImplementation(async (url: string) => {
+      expect(url).toContain("/latest/download/");
+      startDownload();
+      return downloadFinished;
+    }),
+    extractTar: spyOn(tc, "extractTar").mockImplementation(async () => cliDir),
+  };
   const originalArgv1 = process.argv[1];
   process.argv[1] = defaultEntrypoint;
 
   try {
-    await import(`./main.ts?entrypoint=${Date.now()}`);
-    await waitForCalls(() => {
-      expect(spies.setOutput).toHaveBeenCalledWith("version", "supabase 2.84.2");
-      expect(spies.addPath).toHaveBeenCalledWith(cliDir);
+    let importSettled = false;
+    const entrypoint = import(`./main.ts?entrypoint=${Date.now()}`).finally(() => {
+      importSettled = true;
     });
+
+    await downloadStarted;
+    await Bun.sleep(0);
+
+    expect(importSettled).toBe(false);
+
+    finishDownload();
+    await entrypoint;
   } finally {
     process.argv[1] = originalArgv1 ?? "";
   }
 
+  expect(spies.setOutput).toHaveBeenCalledWith("version", "supabase 2.84.2");
+  expect(spies.addPath).toHaveBeenCalledWith(cliDir);
   expect(spies.exportVariable).toHaveBeenCalledWith(CLI_CONFIG_REGISTRY, "ghcr.io");
   expect(spies.setFailed).not.toHaveBeenCalled();
 });

--- a/src/main.ts
+++ b/src/main.ts
@@ -204,5 +204,5 @@ export async function run(): Promise<void> {
 }
 
 if (process.argv[1] === fileURLToPath(import.meta.url)) {
-  void run();
+  await run();
 }

